scaleway/data_source_object_bucket_test.go

```diff
@@ -38,3 +44,191 @@ func TestAccScalewayDataSourceObjectStorage_Basic(t *testing.T) {
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: fakeSideProjectProviders(ctx, tt, project, iamAPIKey),
+		CheckDestroy:      testAccCheckScalewayObjectDestroy(tt),
+		Steps: []resource.TestStep{
+			// Create a bucket from the main provider into the side project and read it from the side provider
+			// The side provider should only be able to read the bucket from the side project
+			{
+				Config: fmt.Sprintf(`
+					resource "scaleway_object_bucket" "base" {
+						name = "%[1]s"
+						project_id = "%[2]s"
+					}
+
+					data "scaleway_object_bucket" "selected" {
+						name = scaleway_object_bucket.base.name
+						provider = side
+					}
+				`,
+					bucketName,
+					project.ID,
+				),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.scaleway_object_bucket.selected", "name", bucketName),
+					resource.TestCheckResourceAttr("data.scaleway_object_bucket.selected", "project_id", project.ID),
+				),
+			},
+		},
+	})
+}
+
+func TestAccScalewayDataSourceObjectStorage_ProjectIDForbidden(t *testing.T) {
+	tt := NewTestTools(t)
+	defer tt.Cleanup()
+	bucketName := sdkacctest.RandomWithPrefix("test-acc-scaleway-object-bucket")
+
+	project, iamAPIKey, terminateFakeSideProject, err := createFakeSideProject(tt)
+	require.NoError(t, err)
+	defer terminateFakeSideProject()
+
+	ctx := context.Background()
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: fakeSideProjectProviders(ctx, tt, project, iamAPIKey),
+		CheckDestroy:      testAccCheckScalewayObjectDestroy(tt),
+		Steps: []resource.TestStep{
+			// The side provider should not be able to read the bucket from the main project
+			{
+				Config: fmt.Sprintf(`
+					resource "scaleway_object_bucket" "base" {
+						name = "%[1]s"
+					}
+
+					data "scaleway_object_bucket" "selected" {
+						name = scaleway_object_bucket.base.name
+						provider = side
+					}
+				`,
+					bucketName,
+					project.ID,
+				),
+				ExpectError: regexp.MustCompile("failed getting Object Storage bucket"),
+			},
+		},
+	})
+}
+
+func createFakeSideProject(tt *TestTools) (*accountV2.Project, *iam.APIKey, func(), error) {
+	t := tt.T
+
+	terminateFunctions := []func(){}
+	terminate := func() {
+		for i := len(terminateFunctions) - 1; i >= 0; i-- {
+			terminateFunctions[i]()
+		}
+	}
+
+	projectName := sdkacctest.RandomWithPrefix("test-acc-scaleway-project")
+	iamApplicationName := sdkacctest.RandomWithPrefix("test-acc-scaleway-iam-app")
+	iamPolicyName := sdkacctest.RandomWithPrefix("test-acc-scaleway-iam-policy")
+
+	projectAPI := accountV2.NewAPI(tt.Meta.scwClient)
+	project, err := projectAPI.CreateProject(&accountV2.CreateProjectRequest{
+		Name: projectName,
+	})
+	if err != nil {
+		terminate()
+		return nil, nil, nil, err
+	}
+	terminateFunctions = append(terminateFunctions, func() {
+		err := projectAPI.DeleteProject(&accountV2.DeleteProjectRequest{
```

Comment:
better add this to the helpers. Could be usefull on the the future. Can you define custom type instead of func()